tools/job-cache/job-cache.cpp

```diff
@@ -0,0 +1,626 @@
+  int fd = open(str, flags);
+  if (fd == -1) {
+    log_fail("open(%s): %s", str, strerror(errno));
+  }
+  return fd;
+}
+
+// Helper that only returns successful file opens and exits
+// otherwise.
+static int open_fd(const char *str, int flags, mode_t mode) {
+  int fd = open(str, flags, mode);
+  if (fd == -1) {
+    log_fail("open(%s): %s", str, strerror(errno));
+  }
+  return fd;
+}
+
+// Like rename but crashes if an error occurs
+static void rename_no_fail(const char *old_path, const char *new_path) {
+  if (rename(old_path, new_path) < 0) {
+    log_fail("rename(%s, %s): %s", old_path, new_path, strerror(errno));
+  }
+}
+
+// Ensures the the given directory has been created
+static void mkdir_no_fail(const char *dir) {
+  if (mkdir(dir, 0777) < 0 && errno != EEXIST) {
+    log_fail("mkdir(%s): %s", dir, strerror(errno));
+  }
+}
+
+// Like close but crashes if an error occurs
+static void close_fd(int fd) {
+  if (close(fd) == -1) {
+    log_fail("close: %s", strerror(errno));
+  }
+}
+
+// This function first attempts to reflink but if that isn't
+// supported by the filesystem it copies instead.
+#ifdef __APPLE__
+
+static void copy_or_reflink(const char *src, const char *dst) {
+  // TODO: Actully make this work. APFS supports reflinking so
+  //       it should be possible to do this correctly
+}
+
+#else
+
+#include <linux/fs.h>
+
+// This function just uses `copy_file_range` to make
+// an efficent copy
+static void copy(int src_fd, int dst_fd) {
+  struct stat buf;
+  if (fstat(src_fd, &buf) < 0) {
+    log_fail("fstat(src_fd = %d): %s", src_fd, strerror(errno));
+  }
+  if (copy_file_range(src_fd, nullptr, dst_fd, nullptr, buf.st_size, 0) < 0) {
+    log_fail(
+        "copy_file_range(src_fd = %d, NULL, dst_fd = %d, size = %d, 0): %s",
+        src_fd, dst_fd, buf.st_size, strerror(errno));
+  }
+}
+
+static void copy_or_reflink(const char *src, const char *dst) {
+  int src_fd = open_fd(src, O_RDONLY);
+  int dst_fd = open_fd(dst, O_WRONLY | O_CREAT, 0644);
+  if (ioctl(dst_fd, FICLONE, src_fd) < 0) {
+    if (errno != EINVAL && errno != EOPNOTSUPP) {
+      log_fail("ioctl(%s, FICLONE, %d): %s", dst, src, strerror(errno));
+    }
+    copy(src_fd, dst_fd);
+  }
+  close_fd(dst_fd);
+  close_fd(src_fd);
+}
+
+#endif
+
+// Use /dev/urandom to get a good seed
+static std::tuple<uint64_t, uint64_t, uint64_t, uint64_t> get_rng_seed() {
+  int rng_fd = open_fd("/dev/urandom", O_RDONLY, 0644);
+  uint8_t seed_data[32] = {0};
+  if (read(rng_fd, seed_data, sizeof(seed_data)) < 0) {
+    log_fail("read(/dev/urandom): %s", strerror(errno));
+  }
+  uint64_t *data = reinterpret_cast<uint64_t *>(seed_data);
+  return std::make_tuple(data[0], data[1], data[2], data[3]);
+}
+
+// Adapted from wikipedia's code, which was adapted from
+// the code included on Sebastiano Vigna's website for
+// Xoshiro256**
+static uint64_t rol64(uint64_t x, int k) { return (x << k) | (x >> (64 - k)); }
+
+static uint8_t hex_to_nibble(char c) {
```

Comment:
what happens when this function is called with 'H' or some other non-hex char?

Reply:
That would be invalid but it would sort of pretend that the numbers kept going higher only it would mod them by 16 so `hex_to_nibble('g') == 0` and `hex_to_nibble('h') == 1`